daemon/daemon.go

```diff
@@ -1032,23 +1036,56 @@ func (daemon *Daemon) TagImage(newTag reference.Named, imageName string, force b
 	return nil
 }
 
+func writeDistributionProgress(cancelFunc func(), outStream io.Writer, progressChan <-chan xfer.Progress) {
+	sf := streamformatter.NewJSONStreamFormatter()
+	for prog := range progressChan {
+		if prog.Message != "" {
+			formatted := sf.FormatStatus(prog.ID, prog.Message)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
+			}
+		} else {
+			jsonProgress := jsonmessage.JSONProgress{Current: prog.Current, Total: prog.Total}
+			formatted := sf.FormatProgress(prog.ID, prog.Action, &jsonProgress)
+			if _, err := outStream.Write(formatted); err != nil {
+				cancelFunc()
+			}
+		}
+	}
+}
+
 // PullImage initiates a pull operation. image is the repository name to pull, and
 // tag may be either empty, or indicate a specific tag to pull.
 func (daemon *Daemon) PullImage(ref reference.Named, metaHeaders map[string][]string, authConfig *cliconfig.AuthConfig, outStream io.Writer) error {
+	// Include a buffer so that slow client connections don't affect
+	// transfer performance.
+	progressChan := make(chan xfer.Progress, 100)
+
+	writesDone := make(chan struct{})
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+
+	go func() {
+		writeDistributionProgress(cancelFunc, outStream, progressChan)
+		close(writesDone)
+	}()
+
 	imagePullConfig := &distribution.ImagePullConfig{
 		MetaHeaders:     metaHeaders,
 		AuthConfig:      authConfig,
-		OutStream:       outStream,
+		ProgressChan:    progressChan,
 		RegistryService: daemon.RegistryService,
 		EventsService:   daemon.EventsService,
 		MetadataStore:   daemon.distributionMetadataStore,
-		LayerStore:      daemon.layerStore,
 		ImageStore:      daemon.imageStore,
 		TagStore:        daemon.tagStore,
-		Pool:            daemon.distributionPool,
+		DownloadManager: daemon.downloadManager,
 	}
 
-	return distribution.Pull(ref, imagePullConfig)
+	err := distribution.Pull(ctx, ref, imagePullConfig)
```

Comment:
How about reusing the context for progress.

```
ctx, cancel := context.WithCancel(context.Background())

if !isTty {
	p, ctx = progress.New(ctx)

	go func() {
		err := p.WriteTo(outStream)
		// or p.Marshal(transformer).WriteTo(outStream)
		if err != nil {
			cancel()
		}
	}
}
// ^ this could even be in postImagesCreate

err := distribution.Pull(ctx, ref, imagePullConfig)
p.Wait() // or writesDone channel
return err


// in progress generator

p = progress.FromContext(ctx)
p.Write(&progress.Item{}) // normally would use helpers instead
```


One thing I don't like about current method is that if we discussed the pull interface cleanup(not in this PR) one idea was that `ImagePullConfig` would not be a place to dump random stuff, but only contain `MetaHeaders` and `AuthConfig` (things that client sends per pull). This heavily reuses the big config struct model again.